src/hmem_cuda.c

```diff
@@ -582,12 +485,49 @@ static int cuda_hmem_verify_devices(void)
 		return -FI_EIO;
 	}
 
-	if (cuda_device_count == 0)
+	FI_INFO(&core_prov, FI_LOG_CORE, "Number of CUDA devices detected: %d\n",
+			cuda_attr.device_count);
+
+	if (cuda_attr.device_count <= 0)
 		return -FI_ENOSYS;
 
 	return FI_SUCCESS;
 }
 
+/*
+ * Loop over pairs of available devices to deduce the overall P2P capabilities
+ */
+static int cuda_hmem_detect_p2p_access_support(void)
+{
+	CUresult cuda_ret;
+	CUdevice dev, peer;
+	int can_access_peer = 1;
+
+	if (cuda_attr.device_count <= 1)
+		return FI_SUCCESS;
+
+	for (dev = 0, peer = 1; peer < cuda_attr.device_count; ++dev, ++peer) {
+		cuda_ret = ofi_cuDeviceCanAccessPeer(&can_access_peer, dev, peer);
+		if (CUDA_SUCCESS != cuda_ret) {
+			FI_WARN(&core_prov, FI_LOG_CORE,
+				"Failed to detect support for peer-to-peer "
+				"access between CUDA devices via "
+				"cuDeviceCanAccessPeer(): %s:%s\n",
+				ofi_cudaGetErrorName(cuda_ret),
+				ofi_cudaGetErrorString(cuda_ret));
+			return -FI_EIO;
+		}
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"Peer access from CUDA device %d -> CUDA device %d : %s\n",
+			dev, peer, can_access_peer ? "Yes" : "No");
+		if (!can_access_peer)
+			return FI_SUCCESS;
+	}
```

Comment:
Went with a simpler non-exhaustive linear algo for now to appease some initialization overhead concerns, but I'm not entirely sure this is the right approach. NVIDIA's [`deviceQuery`](https://github.com/zchee/cuda-sample/blob/master/1_Utilities/deviceQuery/deviceQuery.cpp) reports the peer access support for each permutation, which somewhat indicates it may be possible for peer access to be uni-directional between 2 devices? I've found no evidence online that this is the case, though there's also no evidence that any meaningful amount of overhead from the exhaustive search is incurred.

Reply:
> it may be possible for peer access to be uni-directional between 2 devices?

Follow-up: peer access itself (enabled via `cudaDeviceEnablePeerAccess()` or otherwise) is unidirectional, but I don't think there's any reason something like `cudaDeviceCanAccessPeer()` would evaluate differently based on the order of `{src,dst}Device`.